c/uca/sifter/unisifeggs.cpp

```diff
@@ -0,0 +1,457 @@
+
+class CodePointRange {
+   public:
+    struct iterator {
+        constexpr char32_t operator*() const {
+            return value;
+        }
+
+        constexpr void operator++() {
+            ++value;
+        }
+
+        constexpr std::strong_ordering operator<=>(
+                const iterator& other) const = default;
+
+        char32_t value;
+    };
+
+    static constexpr CodePointRange Inclusive(char32_t first, char32_t last) {
+        return CodePointRange(first, last + 1);
+    }
+
+    constexpr bool empty() const {
+        return first_ >= pastTheEnd_;
+    }
+
+    constexpr bool contains(char32_t c) const {
+        return c >= first_ && c < pastTheEnd_;
+    }
+
+    constexpr CodePointRange intersection(const CodePointRange& other) const {
+        return CodePointRange(std::max(first_, other.first_),
+                              std::min(pastTheEnd_, other.pastTheEnd_));
+    }
+
+    constexpr char32_t front() const {
+        CHECK(!empty());
+        return first_;
+    }
+
+    constexpr char32_t back() const {
+        CHECK(!empty());
+        return pastTheEnd_ - 1;
+    }
+
+    iterator begin() const {
+        return {first_};
+    }
+
+    iterator end() const {
+        return {pastTheEnd_};
+    }
+
+   private:
+    constexpr CodePointRange(char32_t first, char32_t past_the_end)
+        : first_(first), pastTheEnd_(past_the_end) {}
+
+    char32_t first_;
+    char32_t pastTheEnd_;
+};
+
+// TODO(egg): Make this behave like a container.
+class CodePointSet {
+   public:
+    // Empty set.
+    CodePointSet() = default;
+
+    bool contains(char32_t c) const {
+        const auto it = std::partition_point(
+                ranges_.begin(),
+                ranges_.end(),
+                [c](const CodePointRange& range) { return range.back() < c; });
+        return it != ranges_.end() && it->contains(c);
+    }
+
+    void addAll(CodePointRange range) {
+        const auto firstModifiedRange =
+                std::partition_point(ranges_.begin(),
+                                     ranges_.end(),
+                                     [&range](const CodePointRange& r) {
+                                         return r.back() + 1 < range.front();
+                                     });
+        const auto pastModifiedRanges =
+                std::partition_point(ranges_.begin(),
+                                     ranges_.end(),
+                                     [&range](const CodePointRange& r) {
+                                         return r.front() <= range.back() + 1;
+                                     });
+        char32_t insertedRangeFirst = range.front();
+        if (firstModifiedRange != ranges_.end()) {
+            insertedRangeFirst =
+                    std::min(firstModifiedRange->front(), insertedRangeFirst);
+        }
+        char32_t insertedRangeLast = range.back();
+        if (pastModifiedRanges != ranges_.begin()) {
```

Comment:
Should you check against firstModifiedRange and ranges_.end() here?
I don't have a complete image in my head of the possible combinations, but with firstModifiedRange <= pastModifiedRanges <= ranges_.end()
- if firstModifiedRange == pastModifiedRanges then we should not go to pastModifiedRanges - 1
- if pastModifiedRanges == ranges_.end() then we must not read its fields

right??

Reply:
> if firstModifiedRange == pastModifiedRanges then we should not go to pastModifiedRanges - 1

Hmm. I need to think. I also need to come up with a way to have some tests for this thing…

> if pastModifiedRanges == ranges_.end() then we must not read its fields

Indeed, but we don’t: we only ever look at *(pastModifiedRanges - 1).

Reply:
> Hmm. I need to think. I also need to come up with a way to have some tests for this thing…

I did add some tests, but I have yet to think. I will take another look some other day.